### PR TITLE
fix(all-gens): add traded Pokemon EXP bonus (1.5x same-language, 1.7x international)

### LINKED ISSUE #424: bug(all-gens): traded Pokemon EXP bonus (1.5x same-language, 1.7x international) not modeled in ExpContext
## Description
The `ExpContext` interface has no field for whether the gaining Pokemon was obtained via trade (not original trainer). The pokeplatinum decomp shows a 1.5× bonus for same-language trades and 1.7× for international trades. Neither bonus exists in the interface or any implementation.

## Location
- `packages/battle/src/context/types.ts` lines 431–448 — `ExpContext` missing `isTradedPokemon` and `isInternationalTrade` fields
- All gen `calculateExpGain` implementations — no traded Pokemon handling

## Expected vs Actual
- **Expected** (pokeplatinum `battle_script.c` lines 9980–9988): Same-language trade → 1.5×; international trade → 1.7×
- **Actual**: No field for trade status; bonus never calculated

## Severity
MEDIUM — affects any game scenario involving traded Pokemon

## Source
`references/pokeplatinum-main/pokeplatinum-main/src/battle/battle_script.c` lines 9980–9988

---
Filed by Claude Code agent

### PR DESCRIPTION
**Pull request by uehlbran**
## Summary

- Adds `isTradedPokemon?` and `isInternationalTrade?` optional fields to `ExpContext` in `@pokemon-lib-ts/battle`
- Gen 1–2: only the 1.5x same-language bonus applies (no international trade concept in those games)
- Gen 3–9: 1.5x for same-language trades, 1.7x for international trades
- Updates `calculateExpGainClassic()` in `@pokemon-lib-ts/core` with traded-bonus parameters
- Adds inline traded-bonus block to `BaseRuleset.calculateExpGain()` (Gen 7–9)
- Updates Gen 3–6 override `calculateExpGain()` methods accordingly
- Tests: ≥2 cases per gen (not-traded baseline, 1.5x, 1.7x where applicable)

Source: `pret/pokeplatinum src/battle/battle_script.c` lines 9980–9988
  `BattleSystem_PokemonIsOT == FALSE` → traded bonus; `MON_DATA_LANGUAGE != gGameLanguage` → 1.7x (international), else 1.5x

## Test plan

- [x] Gen 1: 1.5x traded bonus applied; `isInternationalTrade` has no effect (Gen 1 has no language metadata)
- [x] Gen 2: 1.5x traded bonus applied; `isInternationalTrade` has no effect (Gen 2 has no language metadata)
- [x] Gen 3: not-traded baseline, 1.5x same-language, 1.7x international, trainer+traded stacking, international > same-language
- [x] Gen 4: 1.5x and 1.7x traded bonus tests
- [x] Gen 5: 1.5x and 1.7x traded bonus tests
- [x] Gen 6: 1.5x and 1.7x traded bonus tests
- [x] Gen 7–9: covered via `BaseRuleset` changes (same inline logic)

## Related Issue

Closes #424

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * EXP calculation updated: traded Pokémon now receive generation-specific bonuses — Gens 1–2 get 1.5×; Gens 3+ get 1.5× for same-language trades and 1.7× for international trades.

* **Tests**
  * Added and extended tests across Gens 1–6 to validate traded/international trade EXP behavior, edge cases, and interactions with existing multipliers.

### CODE REVIEW

File: packages/gen3/tests/exp-traded.test.ts [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+    // Act
+    const result = ruleset.calculateExpGain(context);
+
+    // Assert
+    // floor(floor(floor(64 * 30 / 7) / 1) * 1.5) * 1.5 = floor(411 * 1.5) = 616
+    expect(result).toBe(616);
+  });
+
+  it("given international trade bonus is strictly greater than same-language trade bonus", () => {
```
[the remark this answers is not in the document]
Comment (uehlbran) [reply packages/gen3/tests/exp-traded.test.ts]: Fixed — renamed the test to use proper Given/When/Then framing in commit f1d5e50.

File: packages/core/src/entities/pokemon.ts [hunk 2]
```diff
@@ -72,6 +72,21 @@ export interface PokemonInstance {
...
+   */
+  isTradedPokemon?: boolean;
+
+  /**
+   * Whether the trade was international (different-language cartridge).
+   * Only meaningful when `isTradedPokemon` is true.
+   * Gen 1-2: always false (no language metadata on cartridge).
+   * Source: pret/pokeplatinum src/battle/battle_script.c lines 9980-9988
+   */
+  isInternationalTrade?: boolean;
```
[the remark this answers is not in the document]
Comment (uehlbran) [reply packages/core/src/entities/pokemon.ts]: Fixed in commit 166b417 — both fields are now `readonly` on `PokemonInstance`.

File: packages/gen5/tests/exp-formula.test.ts [hunk 3]
```diff
@@ -189,3 +189,46 @@ describe("Gen5 EXP formula", () => {
...
 });
+
+describe("Gen5Ruleset calculateExpGain — traded Pokemon EXP bonus", () => {
+  const ruleset = new Gen5Ruleset();
+
+  it("given a traded (same-language) Pokemon in Gen 5, when calculateExpGain with isTradedPokemon=true, then returns 1.5x boosted EXP", () => {
+    // Source: Showdown sim/battle-actions.ts — traded bonus applied after other multipliers
+    // Gen 5 sqrt formula: baseExp=100, L_d=50, L_p=50
+    //   a=110, b=110; floor(sqrt(110)*110^2)=126909; exp = floor(126909*100/126909)+1 = 101
+    //   traded (same): floor(101 * 1.5) = 151
```
[the remark this answers is not in the document]
Comment (uehlbran) [reply packages/gen5/tests/exp-formula.test.ts]: Fixed in commit 166b417 — corrected 126909 → 126905 in the comment.
